executor/src/witgen/data_structures/finalizable_data.rs

```diff
@@ -58,84 +129,185 @@ impl<T: FieldElement> FinalizableData<T> {
     pub fn len(&self) -> usize {
-        self.data.len()
+        self.pre_finalized_data.len() + self.finalized_data.len() + self.post_finalized_data.len()
     }
 
     pub fn is_empty(&self) -> bool {
-        self.data.is_empty()
+        self.pre_finalized_data.is_empty()
+            && self.finalized_data.is_empty()
+            && self.post_finalized_data.is_empty()
     }
 
     pub fn push(&mut self, row: Row<T>) {
-        self.data.push(Entry::InProgress(row));
+        if self.finalized_data.is_empty() && self.post_finalized_data.is_empty() {
+            self.pre_finalized_data.push(row);
+        } else {
+            self.post_finalized_data.push(row);
+        }
     }
 
     pub fn pop(&mut self) -> Option<Row<T>> {
-        match self.data.pop() {
-            Some(Entry::InProgress(row)) => Some(row),
-            Some(Entry::Finalized(_)) => panic!("Row already finalized."),
-            None => None,
+        if !self.post_finalized_data.is_empty() {
+            self.post_finalized_data.pop()
+        } else if !self.finalized_data.is_empty() {
+            panic!("Row already finalized");
+        } else {
+            self.pre_finalized_data.pop()
+        }
+    }
+
+    /// Removes the last row, even if it has been finalized.
+    pub fn try_remove_last_row(&mut self) {
+        if !self.post_finalized_data.is_empty() {
+            self.post_finalized_data.pop();
+        } else if !self.finalized_data.is_empty() {
+            self.finalized_data.try_remove_last_row();
+        } else {
+            self.pre_finalized_data.pop();
         }
     }
 
     pub fn extend(&mut self, other: Self) {
-        self.data.extend(other.data);
+        if self.finalized_data.is_empty() {
+            self.pre_finalized_data
+                .append(&mut self.post_finalized_data);
+            self.pre_finalized_data.extend(other.pre_finalized_data);
+            self.finalized_data = other.finalized_data;
+            self.post_finalized_data = other.post_finalized_data;
+        } else if other.finalized_data.is_empty() {
+            self.post_finalized_data.extend(other.pre_finalized_data);
+            self.post_finalized_data.extend(other.post_finalized_data);
+        } else if self.post_finalized_data.is_empty() && other.pre_finalized_data.is_empty() {
+            self.finalized_data.data.extend(other.finalized_data.data);
+            self.post_finalized_data = other.post_finalized_data;
+        } else if other.pre_finalized_data.is_empty() {
+            self.finalize_range(
+                (self.pre_finalized_data.len() + self.finalized_data.len())..self.len(),
+            );
+            self.finalized_data.data.extend(other.finalized_data.data);
+            self.post_finalized_data = other.post_finalized_data;
+        } else {
+            panic!(
+                "Cannot extend. Please try to ensure that `other` does not contain finalized rows."
```

Comment:
could this be an assert at the beginning of the function?

Reply:
Oh and it actually does work if other contain finalized rows. I can change the panic message if you like.